lib/http2/hpack.c

```diff
@@ -154,13 +195,36 @@ static h2o_iovec_t *decode_string(h2o_mem_pool_t *pool, const uint8_t **src, con
         return NULL;
 
     if (is_huffman) {
+        uint8_t hflags = 0;
         if (*src + len > src_end)
             return NULL;
-        if ((ret = decode_huffman(pool, *src, len)) == NULL)
+        if ((ret = decode_huffman(pool, *src, len, &hflags)) == NULL)
             return NULL;
+        /* pseudo-headers are checked later in `decode_header` */
+        if (is_header_name && ret->base[0] != ':') {
+            if (hflags & NGHTTP2_HUFF_INVALID_FOR_HEADER_NAME) {
+                *err_desc = "found an invalid character in header name";
+                return NULL;
+            }
+        }
+        if (hflags & NGHTTP2_HUFF_INVALID_FOR_HEADER_VALUE && !is_header_name) {
+            *err_desc = "found an invalid character in header value";
+            return NULL;
+        }
```

Comment:
How about implementing the check using a single big `if (is_header_name) { ... } else { ... }` block? IMO doing so would improve readability as well as reduce the number of conditional branches.

Also, I think you would need to test the length of the header name before evaluating `ret->base[0]` (and return an invalid header error if the length is zero).